### PR TITLE
tbl_vars() makes a "dplyr_sel_vars" object

### PR DESCRIPTION
**Pull request by romainfrancois**
`tbl_vars.data.frame` makes a `dplyr_sel_vars` object that includes the groups attribute. 

It's not ideal because other implementation of the `tbl_vars` generic would also be expected to do the same thing. 

Maybe instead `tbl_vars` should no longer be generic and be something like: 

```r
tbl_vars <- function(x) {
  vars <- tbl_all_vars(x)
  group_vars <- group_vars(x)

  structure(
    vars,
    class = "dplyr_sel_vars",
    groups = group_vars
  )
}
```

where `tbl_all_vars()` and `group_vars()` are generic. That would also be a pain to synchronize with implementations, i.e. `dbplyr`. 

So we'd have to synchronise the releases and potentially break ?

### CODE REVIEW

**Comment by lionel-** [comment]
Another possibility is to do the dispatch from another function:

```r
tbl_vars_dispatch <- function(x) {
  UseMethod("tbl_vars")
}
```

Then we can do:

```r
tbl_vars <- function(x) {
  vars <- tbl_vars_dispatch(x)
  structure(vars, etc)
}
```

**Comment by romainfrancois** [comment]
Nice trick. I'll have a go and see if it breaks e.g. `dbplyr`

**Comment by hadley** [comment]
Fixes #4106 ?

**Comment by romainfrancois** [comment]
I've implemented @lionel- workaround, which seems to work fine, e.g. 

``` r
library(dplyr, warn.conflicts = FALSE)    

tbl_vars(iris)
#> <dplyr:::vars>
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width" 
#> [5] "Species"
tbl_vars(group_by(iris, Species))
#> <dplyr:::vars>
#> [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width" 
#> [5] "Species"     
#> Groups:
#> [1] "Species"
```

This however causes some `dbplyr` tests to fail, I'll submit a pull request to `dbplyr`: 

```
test-tbl-sql.R:57: failure: can refer to default schema explicitly
tbl_vars(tbl(con, "t1")) not equal to "x".
Attributes: < Modes: list, NULL >
Attributes: < Lengths: 2, 0 >
Attributes: < names for target but not for current >
Attributes: < current is not list-like >
target is dplyr_sel_vars, current is character

test-tbl-sql.R:58: failure: can refer to default schema explicitly
tbl_vars(tbl(con, in_schema("main", "t1"))) not equal to "x".
Attributes: < Modes: list, NULL >
Attributes: < Lengths: 2, 0 >
Attributes: < names for target but not for current >
Attributes: < current is not list-like >
target is dplyr_sel_vars, current is character

test-tbl-sql.R:67: failure: can distinguish 'schema.table' from 'schema'.'table'
tbl_vars(tbl(con, in_schema("aux", "t1"))) not equal to c("x", "y", "z").
Attributes: < Modes: list, NULL >
Attributes: < Lengths: 2, 0 >
Attributes: < names for target but not for current >
Attributes: < current is not list-like >
target is dplyr_sel_vars, current is character

test-tbl-sql.R:68: failure: can distinguish 'schema.table' from 'schema'.'table'
tbl_vars(tbl(con, ident("aux.t1"))) not equal to c("a", "b", "c").
Attributes: < Modes: list, NULL >
Attributes: < Lengths: 2, 0 >
Attributes: < names for target but not for current >
Attributes: < current is not list-like >
target is dplyr_sel_vars, current is character
```

**Comment by romainfrancois** [comment]
@Hadley should I move this to 0.9.0 since https://github.com/tidyverse/dbplyr/pull/312 is not merged in the recently released dbplyr ?

**Comment by hadley** [comment]
Oh shoot sorry. Remind me next week and I’ll do another dbplyr release.

File: R/tbl.r [hunk 1]
```diff
@@ -41,17 +41,34 @@ as.tbl <- function(x, ...) UseMethod("as.tbl")
 #' @export
 as.tbl.tbl <- function(x, ...) x
 
+tbl_vars_dispatch <- function(x) {
+  UseMethod("tbl_vars")
```
Comment (hadley) [inline R/tbl.r]: Wouldn't this be better off just inlined inside of `tbl_vars()`?

Comment (romainfrancois) [reply R/tbl.r]: I'm not sure I understand how to make it work 😬

**Pull request merged**